manager/scheduler/nodeinfo.go

```diff
@@ -94,3 +108,43 @@ func taskReservations(spec api.TaskSpec) (reservations api.Resources) {
 	}
 	return
 }
+
+// taskFailed records a task failure from a given service.
+func (nodeInfo *NodeInfo) taskFailed(ctx context.Context, serviceID string) {
+	if nodeInfo.recentFailures == nil {
+		nodeInfo.recentFailures = make(map[string][]time.Time)
+	}
+
+	expired := 0
+	now := time.Now()
+	for _, timestamp := range nodeInfo.recentFailures[serviceID] {
+		if now.Sub(timestamp) < monitorFailures {
+			break
+		}
+		expired++
+	}
+
+	if len(nodeInfo.recentFailures[serviceID]) == maxFailures-1 {
```

Comment:
I think it should fix recentFailures before checking. 

```
 	nodeInfo.recentFailures[serviceID] = append(nodeInfo.recentFailures[serviceID][expired:], now)

	if len(nodeInfo.recentFailures[serviceID]) == maxFailures {
 		log.G(ctx).Warnf("underweighting node %s for service %s because it experienced %d failures or rejections within %s", nodeInfo.ID, serviceID, maxFailures, monitorFailures.String())
 	}